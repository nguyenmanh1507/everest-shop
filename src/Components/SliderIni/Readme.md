SliderIni example

```js
const data = [
  {
    imgSrc: 'https://unsplash.it/420/560?image=244',
    imgSrcThumbnail: 'https://unsplash.it/90/120?image=244',
    description: 'Description go here'
  },
  {
    imgSrc: 'https://unsplash.it/420/560?image=241',
    imgSrcThumbnail: 'https://unsplash.it/90/120?image=241',
    description: 'Description go here'
  },
  {
    imgSrc: 'https://unsplash.it/420/560?image=242',
    imgSrcThumbnail: 'https://unsplash.it/90/120?image=242',
    description: 'Description go here'
  }
];

<SliderIni data={data} />
```
