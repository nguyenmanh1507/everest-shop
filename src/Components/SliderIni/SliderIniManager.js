// @flow

import * as React from 'react'

type Props = {
  itemsLength: number,
  render: ({
    currentSlide: number,
    setCurrentSlide: (slide: number) => void,
    next: () => void,
    prev: () => void
  }) => React.Node
}

type State = {
  currentSlide: number
}

class SliderIniManger extends React.Component<Props, State> {
  state = {
    currentSlide: 0
  }

  setCurrentSlide = (slide: number) => {
    this.setState({ currentSlide: slide })
  }

  next = () => {
    if (this.state.currentSlide >= this.props.itemsLength - 1) {
      return
    }

    this.setState(({ currentSlide }) => ({
      currentSlide: currentSlide + 1
    }))
  }

  prev = () => {
    if (this.state.currentSlide <= 0) {
      return
    }

    this.setState(({ currentSlide }) => ({
      currentSlide: currentSlide - 1
    }))
  }

  render() {
    const { currentSlide } = this.state
    const { render } = this.props

    return (
      <React.Fragment>
        {render({
          currentSlide,
          setCurrentSlide: this.setCurrentSlide,
          next: this.next,
          prev: this.prev
        })}
      </React.Fragment>
    )
  }
}

export default SliderIniManger
