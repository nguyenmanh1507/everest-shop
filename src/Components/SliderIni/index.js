// @flow

export { default } from './SliderIni'
