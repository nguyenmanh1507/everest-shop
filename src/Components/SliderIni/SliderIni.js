// @flow

import React, { Component } from 'react'
import cx from 'classnames'

import SliderIniManager from './SliderIniManager'

type Props = {
  /** Slider data */
  data: Array<{
    /** Image source */
    imgSrc: string,
    /** Image thumbnail source */
    imgSrcThumbnail: string,
    /** Image description source */
    description: string
  }>
}

class SliderIni extends Component<Props> {
  render() {
    const { data } = this.props

    return (
      <SliderIniManager
        itemsLength={data.length}
        render={({ currentSlide, setCurrentSlide, prev, next }) => (
          <div className="eps">
            <div className="eps__content">
              <img
                data-test="eps-content-img"
                src={data[currentSlide].imgSrc}
                alt={data[currentSlide].description}
              />
              <button className="eps__cta">
                <i className="fa fa-arrows-alt" />
              </button>
            </div>
            <div className="eps__list-thumbnail">
              {data.map((s, index) => (
                <div
                  data-test="eps-thumbnail"
                  key={`sliderini-thumbnail-${index}`}
                  className={cx('eps__thumbnail', {
                    active: currentSlide === index
                  })}
                  onClick={() => {
                    setCurrentSlide(index)
                  }}
                >
                  <img
                    src={s.imgSrcThumbnail}
                    className="eps__thumbnail-img"
                    alt={s.description}
                  />
                </div>
              ))}

              <div className="eps__paginate">
                <button className="eps__paginate-btn" onClick={prev}>
                  <i className="fa fa-angle-up" />
                </button>
                <button className="eps__paginate-btn" onClick={next}>
                  <i className="fa fa-angle-down" />
                </button>
              </div>
            </div>
          </div>
        )}
      />
    )
  }
}

export default SliderIni
