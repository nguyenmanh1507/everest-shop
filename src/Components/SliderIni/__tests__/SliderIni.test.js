// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { render, configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import SliderIni from '../SliderIni'

configure({ adapter: new Adapter() })

test('SliderIni render properly', () => {
  const component = renderer.create(renderSliderIni())

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

describe('<SliderIni>', () => {
  test('After click, thumbnail should has active class', () => {
    const wrapper = mount(renderSliderIni())

    // select second thumbnail
    let thumbnail = wrapper.find('[data-test="eps-thumbnail"]').at(1)
    expect(thumbnail.hasClass('active')).toBeFalsy()
    // click thumbnail
    thumbnail.simulate('click')
    // select again
    thumbnail = wrapper.find('[data-test="eps-thumbnail"]').at(1)
    expect(thumbnail.hasClass('active')).toBeTruthy()
  })

  // TODO: test next, prev function
})

function renderSliderIni() {
  const data = [
    {
      imgSrc: 'https://unsplash.it/420/560?image=244',
      imgSrcThumbnail: 'https://unsplash.it/90/120?image=244',
      description: 'Description go here'
    },
    {
      imgSrc: 'https://unsplash.it/420/560?image=241',
      imgSrcThumbnail: 'https://unsplash.it/90/120?image=241',
      description: 'Description go here'
    },
    {
      imgSrc: 'https://unsplash.it/420/560?image=242',
      imgSrcThumbnail: 'https://unsplash.it/90/120?image=242',
      description: 'Description go here'
    }
  ]

  return <SliderIni data={data} />
}
