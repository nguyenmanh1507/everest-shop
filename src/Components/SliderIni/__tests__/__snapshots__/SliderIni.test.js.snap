// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderIni render properly 1`] = `
<div
  className="eps"
>
  <div
    className="eps__content"
  >
    <img
      alt="Description go here"
      data-test="eps-content-img"
      src="https://unsplash.it/420/560?image=244"
    />
    <button
      className="eps__cta"
    >
      <i
        className="fa fa-arrows-alt"
      />
    </button>
  </div>
  <div
    className="eps__list-thumbnail"
  >
    <div
      className="eps__thumbnail active"
      data-test="eps-thumbnail"
      onClick={[Function]}
    >
      <img
        alt="Description go here"
        className="eps__thumbnail-img"
        src="https://unsplash.it/90/120?image=244"
      />
    </div>
    <div
      className="eps__thumbnail"
      data-test="eps-thumbnail"
      onClick={[Function]}
    >
      <img
        alt="Description go here"
        className="eps__thumbnail-img"
        src="https://unsplash.it/90/120?image=241"
      />
    </div>
    <div
      className="eps__thumbnail"
      data-test="eps-thumbnail"
      onClick={[Function]}
    >
      <img
        alt="Description go here"
        className="eps__thumbnail-img"
        src="https://unsplash.it/90/120?image=242"
      />
    </div>
    <div
      className="eps__paginate"
    >
      <button
        className="eps__paginate-btn"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-up"
        />
      </button>
      <button
        className="eps__paginate-btn"
        onClick={[Function]}
      >
        <i
          className="fa fa-angle-down"
        />
      </button>
    </div>
  </div>
</div>
`;
